chips/virtio/src/devices/virtio_input.rs

```diff
@@ -0,0 +1,174 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2025.
+
+//! Support for the VirtIO Input Device
+//!
+//! <https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-3850008>
+//!
+//! This implementation assumes the input device is a keyboard that follows the
+//! [event code](https://www.kernel.org/doc/Documentation/input/event-codes.txt)
+//! format that Linux uses.
+
+use kernel::utilities::cells::OptionalCell;
+
+use crate::devices::{VirtIODeviceDriver, VirtIODeviceType};
+use crate::queues::split_queue::{SplitVirtqueue, SplitVirtqueueClient, VirtqueueBuffer};
+
+/// Event: separate events.
+const EV_SYN: u16 = 0;
+/// Event: state change of keyboard.
+const EV_KEY: u16 = 1;
+
+/// VirtIO for input devices (e.g., keyboards).
+pub struct VirtIOInput<'a> {
+    /// Queue of events from the device (e.g., keyboard).
+    eventq: &'a SplitVirtqueue<'static, 'static, 3>,
+    /// Queue of status updates from this driver (e.g., turn on LED).
+    statusq: &'a SplitVirtqueue<'static, 'static, 1>,
+    /// Buffer to hold status updates.
+    status_buffer: OptionalCell<&'static mut [u8]>,
+    /// Store keys sent across multiple events.
+    keys: [OptionalCell<(u16, bool)>; 2],
+    /// Keyboard callback client.
+    client: OptionalCell<&'a dyn kernel::hil::keyboard::KeyboardClient>,
+}
+
+impl<'a> VirtIOInput<'a> {
+    pub fn new(
+        eventq: &'a SplitVirtqueue<'static, 'static, 3>,
+        statusq: &'a SplitVirtqueue<'static, 'static, 1>,
+        status_buffer: &'static mut [u8],
+    ) -> Self {
+        eventq.enable_used_callbacks();
+
+        Self {
+            eventq,
+            statusq,
+            status_buffer: OptionalCell::new(status_buffer),
+            keys: [const { OptionalCell::empty() }; 2],
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn provide_buffers(
```

Comment:
Should probably just be called `initialize`. I'd also put the buffers in the constructor, but that would require another `TakeCell<(buf1, buf2, buf3)>`, so it seems fine to accept them here.

And I'd add a sanity check making sure the device isn't already initialized. Otherwise, presumably, you'd get an error exceeding the VirtQueue's capacity.

Reply:
This seems okay as-is to me.